### DISCUSSION TITLE
Looking for assistance with downloading Java

### DISCUSSION THREAD

**Discussion by H12900** [Audiveris/audiveris#745] [General]
Hello,

I understand you need Java to use Audiveris, but I'm having issues with downloading it. I just don't know how. I'm at this page: https://www.oracle.com/java/technologies/downloads/#jdk17-windows, and I'm wondering if there's a difference in the downloads. The first option (x64 Compressed Archive) I have been trying, and it comes as a zipped folder, and I have to do a bunch of stuff in System Properties, etc. etc. But, before that I accidentally downloaded Java version 8, and it was such a simple task, but downloading Java 17 seems impossible!

Now, after all my fancy work in System Preferences and Environment Variables and all that, I still get "'java' is not recognized as an internal or external command, operable program or batch file." when running the "java -version" command.

Any advice or assistance would be GREATLY appreciated.

**Comment by hbitteur** [comment]
@H12900 
I infer from your message (please correct me if I'm wrong) that:

1. You are running under Windows
2. You have used the Audiveris Windows installer to install the Audiveris software. This installer takes care of many things, except the needed Java environment which you have to handle on your own.
3. You tried to download and install Java 17 but failed to do so.

From the same web page mentioned in your message, don't waste your time fighting with the x64 compressed archive, simply select the "**x64 Installer**" which will take care of everything (just download the installer and run it).

Then use the command "java -version" to make sure you are now using Java 17.

For more detailed information you can read the Oracle [JDK installation documentation](https://docs.oracle.com/en/java/javase/17/install/installation-jdk-microsoft-windows-platforms.html#GUID-A7E27B90-A28D-4237-9383-A58B416071CA).

**Comment by H12900** [comment]
Thank you for your reply.

Yes, all three are correct. But i have now installed Java 17 correctly (I think) and when i ran "java -version" I got "java version "17.0.12" 2024-07-16 LTS
Java(TM) SE Runtime Environment (build 17.0.12+8-LTS-286)
Java HotSpot(TM) 64-Bit Server VM (build 17.0.12+8-LTS-286, mixed mode, sharing)"

However, whatever I put into the Command Center to launch Audiveris, nothing works to launch it. I always get some sort of error.

(Later)

I moved the Executable Jar File (.jar) with the name "Audiveris" to a new location, and finally got something different than the error messages (when entering its address into the Command Center), and that was a notice that my PC can't run the program. Is it possible that the firewall my computer is protected by is not allowing this?

**Reply by hbitteur** [reply]
> whatever I put into the Command Center to launch Audiveris

Sorry, what do you mean by "Command Center"?

The standard Audiveris installation on Windows results in a new item in the Windows start menu (accessible via an icon similar to a window, located in the lower left side of the desktop).
This start menu provides a list of applications, sorted alphabetically. There, it's easy to find the Audiveris item.
Please have a look at [installing binaries](https://audiveris.github.io/audiveris/_pages/install/binaries/) documentation.

**Reply by hbitteur** [reply]
> I moved the Executable Jar File (.jar) with the name "Audiveris" to a new location

No, don't do that.
Audiveris is launched via a command file (named Audiveris.bat, generally located at c:\Program files\Audiveris\bin) which launches java.exe with proper parameters, including the location of Audiveris.jar.

**Reply by hbitteur** [reply]
> However, whatever I put into the Command Center to launch Audiveris, nothing works to launch it. I always get some sort of error.

Please post the precise action you make and the messages you get, so that we can help you effectively.

**Comment by H12900** [comment]
I think that was supposed to be "Command Prompt" in Windows, where you give it a prompt or a command. 

So my goal is to launch Audiveris. I have tried the following steps:

Double click on the folder you mentioned, which causes a window to flash quickly but nothing else happens. Also tried open with (see img)
![Picture1](https://github.com/user-attachments/assets/db1ed9f3-69a4-4637-8d6e-5b5b0b6bcc25)

As I mentioned, I'm not quite sure I'm doing the correct steps to launch Audiveris, so i had an AI [Poe](https://poe.com/) guide me through the process. He told me to do the first step (which I mentioned didn't work above) and then he said to "use the command line to run the Audiveris application by executing the following command:" "C:\Program Files\Audiveris\bin\audiveris.exe"

Here are the fruits of that: 
![Picture2](https://github.com/user-attachments/assets/30ee2dcd-da75-47b7-92fa-1947e1819cc0)

So, I am stuck. Why can't I open the program? I am willing to answer any other questions you guys might have, and I hope that the information I have provided has been helpful. 

Thank you.

**Comment by hbitteur** [comment]
Here are 3 solutions, pick up the one you prefer.

1/ Why not simply using the Windows start menu?
- Click on Windows start icon (bottom left of your screen)
- Navigate to Audiveris submenu
- Select the first item (Audiveris) not the second one (Website)

2/ In the image you posted, the File Explorer presents the content of `C:\Program Files\Audiveris\lib`. Please read carefully: this is `lib`, not `bin`! Please move to  `C:\Program Files\Audiveris\bin` instead, and there you can double-click on `Audiveris.bat`.

3/ In the second image (command prompt window), you typed: "C:\Program Files\Audiveris\bin\audiveris.exe". Here again read carefully: this is `.exe` not `.bat`

**Comment by H12900** [comment]
Ok,

When I scrolled through the Start menu, I saw absolutely nothing referring to Audiveris. 
![Start Menu](https://github.com/user-attachments/assets/ff57de29-8819-406d-b5f8-082c4dede4de)

Then, I tried opening the file you specified by double-clicking on it. The screen just flashed, as if it were trying to open it but couldn't. I also tried right-clicking on the file, and clicking on "Run as Administrator." That didn't work either.
![Picture3](https://github.com/user-attachments/assets/6140c5bb-855c-417c-a4a7-7735a10340d4)

Lastly, I tried running the command you specified in the Command Prompt. That yielded no results either:
![Command Prompt](https://github.com/user-attachments/assets/fcfcbf23-2daf-4d18-a1c1-1113cbc7f74e)

Thank you for your patience and helpfulness.

**Comment by hbitteur** [comment]
So, your JAVA_HOME environment variable points to `c:\Users\Hosea\Downloads\jdk-17_windows-x64_bin`
A strange location.
As an example, mine points to `c:\Program Files\Java\jdk-21` (I'm using Java 21 for development).

Plus the fact that your start menu does not contain any Audiveris submenu.
This makes me think that you have not installed Audiveris and Java correctly.
As if you had simply expanded a Java compressed archive rather than launching the installer.

Make sure you go to [this web page](https://www.oracle.com/java/technologies/downloads/#jdk17-windows) to download and run the "x64 Installer", not the "x64 Compressed Archive".

**Comment by H12900** [comment]
Ok,

I deleted/uninstalled, and reinstalled Java and Audiveris, and Audiveris still doesn't launch. At the start when it installed Audiveris, it said: System prevented an unknown app from running", something like that. But I just pressed "Run anyways." 

Double clicking the .exe file still causes the screen to flash, clicking "Run as administrator" doesn't work either, and trying to launch from Command Prompt is also fruitless. 

![Command Prompt 2](https://github.com/user-attachments/assets/c71e399a-7214-4e59-8c07-9a0269c22196)

![Java Version](https://github.com/user-attachments/assets/9f527ffd-b550-4c19-a71d-f5b9489268d7)

![Java Location](https://github.com/user-attachments/assets/ec0fc3f0-c454-4f46-a5d2-3bafe1339742)

**Comment by hbitteur** [comment]
According to your second image (java -version), you now seem to have Java 17 up and running. This is a good starting point.

According to your first image (audiveris.exe launched from c:\Program Files\Audiveris\bin folder), audiveris.exe is not recognized, etc...
But there must be no such `audiveris.exe` file in this folder! Could you instead launch `Audiveris.bat`? Pay attention, this is **.bat** extension instead of .exe extension.

**Comment by hbitteur** [comment]
Could you post an image presenting the content of your `c:\Program Files\Audiveris\bin` folder

**Comment by H12900** [comment]
Sure,

![bin](https://github.com/user-attachments/assets/a784fb0b-44c7-437b-8c7d-9180a78ae2f6)

**Comment by hbitteur** [comment]
First line is `Audiveris` (w/ no extension). It's meant for sh or bash (this is not your case).
Second line is `Audiveris.bat`, meant for Windows. This is the one you should use to launch Audiveris.
Third line is `icon-256.ico`. This is the icon automatically used by Windows when displaying the Audiveris application.

Note there is no `Audiveris.exe` file.

You should configure your File Explorer so that it displays the extension of every file (use the "View" tab and there select the item "File name extensions". This will make the files view easier to read.

**Comment by H12900** [comment]
Ok, I configured File Explorer as you suggested.

When I try to open Audiveris.bat a screen flashes; it almost looks like the Command Prompt screen, but other than that nothing happens.

I'm not sure what do do about the missing Audiveris.exe file.

**Comment by hbitteur** [comment]
To be able to see any error message before the windows closes, you have to:
1/ Open a Command Prompt window
2/ From this window, call the Audiveris.bat file.
Then read the messages in the window (and post them here).
I strongly suspect you don't have the proper Java environment correctly installed.

Could you tell us which version of Audiveris installer you have used ([Audiveris_Setup-5.3.1-windows-x86_64.exe](https://github.com/Audiveris/audiveris/releases/download/5.3.1/Audiveris_Setup-5.3.1-windows-x86_64.exe)?).
If installed correctly, you should have an Audiveris sub-menu in Windows start menu.

And there is **NO** Audiveris.exe,  don't look for it.
The Audiveris.bat file calls java.exe with Audiveris.jar and other needed libraries.

**Comment by H12900** [comment]
I just checked again, the sub-menu is now in the Start Menu; what has made it appear, I don't know. When I left-clicked on the first option, (as you instructed previously,) a command prompt window opened for a split second, and it said "Audiveris" at the top, but then it vanished.

Then I entered the path to the Audiveris.bat file, and after pressing enter, I typed Audiveris.bat, and here here is the result:

![Error Message](https://github.com/user-attachments/assets/03ead58c-d130-4c31-8103-e153449fdb2d)

Error Message: ERROR: JAVA_HOME is set to an invalid directory: C:\Users\Hosea\Downloads\jdk-17_windows-x64_bin

Please set the JAVA_HOME variable in your environment to match the
location of your Java installation.

**Comment by hbitteur** [comment]
But we already discussed this error message last week.
I told you this:
> So, your JAVA_HOME environment variable points to c:\Users\Hosea\Downloads\jdk-17_windows-x64_bin
A strange location.
As an example, mine points to c:\Program Files\Java\jdk-21 (I'm using Java 21 for development).

Please make sure your environment variable JAVA_HOME points to your Java installation, probably `c:\Program Files\Java\jdk-17` (and certainly not to some download folder)!
What have you done to correct this? We are wasting time.

**Comment by H12900** [comment]
Well, perhaps if you had given me specific instructions last week, as you did now, we would not have wasted any time. I'm new to this stuff, and cannot act to fix this issue without being told what to do. 

I have set the environment variable as you instructed me to, and Audiveris successfully launched. 

I thank you kindly for your assistance.

**Comment by hbitteur** [comment]
I'm glad you finally succeeded.

For your information, in the not yet released version 5.4, the Java environment is checked before Audiveris actually starts and the launch window remains open if an error is detected (allowing you to read the messages).
This new feature would have saved us a week...
